core/primitives/src/sharding.rs

```diff
@@ -606,12 +609,52 @@ impl ShardChunkHeader {
                 // In bandwidth scheduler version v3 and v4 are allowed. The first chunk in
                 // the bandwidth scheduler version will be v3 because the chunk extra for the
                 // last chunk of previous version doesn't have bandwidth requests.
-                ShardChunkHeaderInner::V2(_) => {
-                    version >= BLOCK_HEADER_V3_VERSION && version < BANDWIDTH_SCHEDULER_VERSION
-                }
+                // v2 is also allowed in the bandwidth scheduler version because there
+                // are multiple tests which upgrade from an old version directtly to the
+                // latest version. TODO(#12328) - don't allow InnerV2 in bandwidth scheduler version.
+                ShardChunkHeaderInner::V2(_) => version >= BLOCK_HEADER_V3_VERSION,
                 ShardChunkHeaderInner::V3(_) => version >= CONGESTION_CONTROL_VERSION,
                 ShardChunkHeaderInner::V4(_) => version >= BANDWIDTH_SCHEDULER_VERSION,
             },
+        };
+
+        if is_valid {
+            Ok(())
+        } else {
+            Err(BadHeaderForProtocolVersionError {
+                protocol_version: version,
+                header_version: self.header_version_number(),
+                header_inner_version: self.inner_version_number(),
+            })
+        }
+    }
+
+    /// Used for error messages, use `match` for other code.
+    #[inline]
+    pub(crate) fn header_version_number(&self) -> u64 {
+        match self {
+            ShardChunkHeader::V1(_) => 1,
+            ShardChunkHeader::V2(_) => 2,
+            ShardChunkHeader::V3(_) => 3,
+        }
+    }
+
+    /// Used for error messages, use `match` for other code.
+    #[inline]
+    pub(crate) fn inner_version_number(&self) -> u64 {
+        match self {
+            ShardChunkHeader::V1(v1) => {
+                let _inner_v1: &ShardChunkHeaderInnerV1 = &v1.inner;
```

Comment:
why do you need this unused variable?

Reply:
It "proves" that both `ShardChunkHeader::V1` and `ShardChunkHeader::V2` contain `ShardChunkHeader::V1`. It can be confusing that for headerv2 we have to return 1 instead of 2. I wanted to make this more readable/reviewable.

Reply:
Ah cool but can you also add a comment please? :)